File: Scripts/scaling.py
import numpy as np
import scipy.sparse as sp
import time

# ── Neumann series approximation of (I - A)^{-1} v ─────────────────────────
def neumann_apply(A, v, k=20):
    """Compute (I + A + A^2 + ... + A^k) v via Horner's method."""
    # Ensure v is a dense array for efficient sparse @ dense
    v = np.asarray(v)
    result = v.copy()
    term   = v.copy()
    for _ in range(k):
        term   = A @ term
        result = result + term
    return result

# ── Main solver ──────────────────────────────────────────────────────────────
def solve_planner(alpha, A_bar, B, l_tilde, dK, K, L_total, G_vec, gamma=None,
                  C_prev=None,
                  k=20,
                  tol=1e-4,
                  eta_K=0.4, eta_L=0.4,
                  max_iter=2000):

    eps_val = 1e-15
    n       = len(alpha)

    # Default LES subsistence thresholds to zero if not provided
    if gamma is None:
        gamma = np.zeros(n)

    # B operator  Bt(v)  = B * (I-A)^{-1} v
    # Adjoint     BtT(w) = (I-A)^{-T} * B^T w   [B diagonal => B^T = B]
    def Bt(v):
        return B * neumann_apply(A_bar, v, k)

    # Cache transpose for efficiency in sparse context
    A_bar_T = A_bar.transpose()
    def BtT(w):
        return neumann_apply(A_bar_T, B * w, k)

    # Effective resources net of committed investment
    K_eff  = K - Bt(dK)
    L_eff  = L_total - np.dot(l_tilde, dK)

    active     = K_eff > eps_val
    Kv         = K_eff[active]
    alpha_act  = alpha[active]
    gamma_act  = gamma[active]
    l_act      = l_tilde[active]
    n_act      = int(active.sum())

    def Bt_act(v_act):
        v_full = np.zeros(n); v_full[active] = v_act
        return Bt(v_full)[active]

    def BtT_act(w_act):
        w_full = np.zeros(n); w_full[active] = w_act
        return BtT(w_full)[active]

    # Cold / warm start
    if C_prev is not None:
        p_guess = alpha_act / np.maximum(C_prev[active] - gamma_act, eps_val)
        w_guess = np.zeros(n)
        w_guess[active] = p_guess
        w_res   = w_guess - A_bar_T @ w_guess
        lam_K   = np.maximum(w_res[active], eps_val)
    else:
        lam_K = np.full(n_act, (alpha_act.sum() / n_act) / max((Kv.sum() / n_act), eps_val))
        lam_K = np.maximum(lam_K, eps_val)

    lam_L = (alpha_act.sum() / n_act) / max(L_eff, eps_val)
    
    log_lam_K = np.log(lam_K)
    log_lam_L = np.log(lam_L)

    log_y_K = log_lam_K.copy()
    log_y_L = log_lam_L

    prev_grad_K   = np.zeros(n_act)
    prev_grad_L   = 0.0
    prev_dtheta_K = np.zeros(n_act)
    prev_dtheta_L = 0.0

    eta_K_cur = eta_K
    eta_L_cur = eta_L

    C_act     = np.zeros(n_act)
    converged = False
    opt_iter  = 0

    for it in range(1, max_iter + 1):
        opt_iter = it

        y_K = np.exp(log_y_K)
        y_L = np.exp(log_y_L)
        
        # LES stationarity -> closed-form C = gamma + alpha / pi
        pi_vec = BtT_act(y_K) + y_L * l_act
        C_act  = gamma_act + alpha_act / np.maximum(pi_vec, eps_val)

        # Slacks
        s_K = Kv   - Bt_act(C_act)
        s_L = L_eff - np.dot(l_act, C_act)

        # KKT check
        primal_ok_K = np.all(s_K / (Kv + eps_val) >= -tol)
        primal_ok_L = (s_L / (abs(L_eff) + eps_val)) >= -tol
        lam_K_cur = np.exp(log_lam_K)
        lam_L_cur = np.exp(log_lam_L)
        comp_ok_K   = np.all(np.abs(lam_K_cur * s_K) <= tol)
        comp_ok_L   = abs(lam_L_cur * s_L) <= tol

        if primal_ok_K and primal_ok_L and comp_ok_K and comp_ok_L:
            converged = True
            break

        grad_K = (-s_K) / (Kv + eps_val)
        grad_L = (-s_L) / (abs(L_eff) + eps_val)

        # Barzilai-Borwein adaptive step
        if it > 1:
            dy_K = grad_K - prev_grad_K
            dot_ss = np.dot(prev_dtheta_K, prev_dtheta_K) + prev_dtheta_L**2
            dot_sg = np.dot(prev_dtheta_K, dy_K) + prev_dtheta_L * (grad_L - prev_grad_L)
            if abs(dot_sg) > 1e-30:
                bb_eta = np.clip(abs(dot_ss / dot_sg), 0.01, 0.5)
                eta_K_cur = bb_eta
                eta_L_cur = bb_eta

        prev_grad_K = grad_K.copy()
        prev_grad_L = grad_L

        step_K = eta_K_cur * grad_K
        step_L = eta_L_cur * grad_L

        prev_dtheta_K = step_K.copy()
        prev_dtheta_L = step_L

        log_lam_K_new = log_y_K + step_K
        log_lam_L_new = log_y_L + step_L

        # Nesterov momentum update
        beta_t = (it - 1.0) / (it + 2.0)
        log_y_K = log_lam_K_new + beta_t * (log_lam_K_new - log_lam_K)
        log_y_L = log_lam_L_new + beta_t * (log_lam_L_new - log_lam_L)

        log_lam_K = log_lam_K_new.copy()
        log_lam_L = log_lam_L_new
        
    lam_K = np.exp(log_lam_K)
    lam_L = np.exp(log_lam_L)

    # Recovery
    lam_K_full          = np.zeros(n)
    lam_K_full[active]  = lam_K
    pi_star             = BtT(lam_K_full) + lam_L * l_tilde

    C_star          = np.zeros(n)
    C_star[active]  = C_act
    X_star          = neumann_apply(A_bar, C_star + dK + G_vec, k)

    return dict(C_star=C_star, X_star=X_star, pi_star=pi_star,
                success=converged, lambda_K=lam_K_full, lambda_L=lam_L,
                iterations=opt_iter)


# ── Synthetic data generator ─────────────────────────────────────────────────
def make_synthetic(n, sparsity=0.15, seed=42):
    rng = np.random.default_rng(seed)

    # Sparse A with spectral radius < 0.65
    mask = rng.random((n, n)) < sparsity
    A_raw = rng.uniform(0, 1, (n, n)) * mask
    A_raw_sp = sp.csr_matrix(A_raw)
    
    # Sparse spectral radius estimation
    if n > 1:
        vals = sp.linalg.eigs(A_raw_sp, k=1, which='LM', return_eigenvectors=False)
        sr = np.abs(vals[0])
    else:
        sr = A_raw[0,0]
        
    A = A_raw_sp * (0.60 / max(sr, 1e-10))

    # Depreciation-augmented matrix
    delta = 0.0125
    B     = rng.uniform(0.05, 0.3, n)    # diagonal capital coefficients
    
    # Create sparse A_bar: CSR is efficient for matrix-vector multiplication
    A = sp.csr_matrix(A)
    A_bar = A + delta * sp.diags(B)

    alpha  = rng.dirichlet(np.ones(n))
    l_vec  = rng.uniform(0.1, 0.5, n)
    L_total = n * 0.3                    # aggregate labour endowment

    # l_tilde = (I - A_bar)^{-T} l
    # For small synthetic n, we can use sparse solve, but for benchmark 
    # we usually hold l_tilde as dense.
    I_sparse = sp.eye(n, format='csr')
    l_tilde = sp.linalg.spsolve((I_sparse - A_bar).T, l_vec)

    # Capital stock: K = B * X_0, X_0 ~ uniform
    X0  = rng.uniform(1, 3, n)
    K   = B * X0 * 1.2                  # 20% buffer

    dK  = rng.uniform(0.01, 0.05, n)
    G   = rng.uniform(0.05, 0.15, n)

    return dict(alpha=alpha, A_bar=A_bar, B=B, l_tilde=l_tilde,
                dK=dK, K=K, L_total=L_total, G_vec=G)


# ── Benchmark ────────────────────────────────────────────────────────────────
def run_benchmark(n, n_trials=5):
    print(f"\n{'='*55}")
    print(f"  Benchmark: n = {n} sectors, {n_trials} trials")
    print(f"{'='*55}")

    data = make_synthetic(n)
    times      = []
    iterations = []
    successes  = []

    C_prev = None
    for t in range(n_trials):
        t0  = time.perf_counter()
        res = solve_planner(
            data['alpha'], data['A_bar'], data['B'],
            data['l_tilde'], data['dK'], data['K'],
            data['L_total'], data['G_vec'],
            C_prev=C_prev
        )
        elapsed = time.perf_counter() - t0
        times.append(elapsed)
        iterations.append(res['iterations'])
        successes.append(res['success'])
        C_prev = res['C_star']   # warm start next trial

    print(f"\n  {'Trial':<8} {'Time (s)':<12} {'Iterations':<12} {'Converged'}")
    print(f"  {'-'*45}")
    for i, (t, it, s) in enumerate(zip(times, iterations, successes)):
        warm = " (warm)" if i > 0 else " (cold)"
        print(f"  {i+1:<8} {t:<12.4f} {it:<12} {s}{warm}")

    print(f"\n  Mean time      : {np.mean(times):.4f}s")
    print(f"  Min  time      : {np.min(times):.4f}s")
    print(f"  Mean iterations: {np.mean(iterations):.1f}")
    print(f"  Min  iterations: {np.min(iterations)}")
    print(f"  All converged  : {all(successes)}")

    # Verify material balance
    r = res
    I_sparse = sp.eye(n, format='csr')
    net_output = (I_sparse - data['A_bar']) @ r['X_star']
    final_demand = r['C_star'] + data['dK'] + data['G_vec']
    balance_err  = np.max(np.abs(net_output - final_demand))
    print(f"  Material balance error : {balance_err:.2e}")

    return times, iterations


if __name__ == "__main__":
    run_benchmark(2000, n_trials=3)

File: Scripts/scaling.jl
using LinearAlgebra
using SparseArrays
using Random
using Printf

# ── Neumann series approximation of (I - A)^{-1} v ─────────────────────────
mutable struct NeumannCache
    result::Vector{Float64}
    term::Vector{Float64}
    term_next::Vector{Float64}
    NeumannCache(n::Int) = new(zeros(n), zeros(n), zeros(n))
end

function neumann_apply!(cache::NeumannCache, A_bar, v, k::Int=20)
    copyto!(cache.result, v)
    copyto!(cache.term, v)
    for _ in 1:k
        mul!(cache.term_next, A_bar, cache.term)
        cache.result .+= cache.term_next
        cache.term, cache.term_next = cache.term_next, cache.term
    end
    return cache.result
end

# ── Main solver ──────────────────────────────────────────────────────────────
# --- CRRA Dual Solver (FISTA + Backtracking) ---
function solve_planner(alpha_v, A_m, B_m, l_tilde_v, dK_v, K_v, L_total_f, G_v, sigma_v, C_prev_v;
                       k::Int=25, tol_p::Float64=1e-3, tol_d::Float64=1e-4, 
                       eta_K::Float64=0.4, eta_L::Float64=0.5, max_iter::Int=2000,
                       L0::Float64=1.0, L_scale_up::Float64=1.5, L_scale_dn::Float64=1.05)

    eps_val = 1e-15
    n = length(alpha_v)
    cache = NeumannCache(n)
    A_m_T = sparse(A_m')

    Bt(v)  = B_m .* neumann_apply!(cache, A_m, v, k)
    BtT(w) = neumann_apply!(cache, A_m_T, B_m .* w, k)

    K_eff = K_v .- Bt(dK_v)
    L_eff = L_total_f - dot(l_tilde_v, dK_v)

    # Initialization
    pi_init = alpha_v ./ max.(C_prev_v, 1e-4).^sigma_v
    log_lam_K = log.(max.(BtT(pi_init), 1e-8))
    log_lam_L = log(sum(alpha_v) / max(n * L_eff, eps_val))

    log_y_K = copy(log_lam_K); log_y_L = log_lam_L
    log_x_K = copy(log_lam_K); log_x_L = log_lam_L
    log_x_K_new = copy(log_lam_K)
    
    tk_curr = 1.0; L_curr = L0
    converged = false; opt_iter = max_iter

    for iter in 1:max_iter
        opt_iter = iter
        y_K = exp.(log_y_K); y_L = exp(log_y_L)
        pi_vec = BtT(y_K) .+ y_L .* l_tilde_v
        C_res = (max.(alpha_v, 1e-15) ./ max.(pi_vec, eps_val)).^(1.0 ./ sigma_v)

        s_K = K_eff .- Bt(C_res)
        s_L = L_eff  - dot(l_tilde_v, C_res)

        # Grad check
        grad_K = .-s_K ./ max.(K_eff, 1e-12)
        grad_L = -s_L / max(abs(L_total_f), 1e-12)

        if all(abs.(y_K .* s_K) .<= tol_d) && abs(y_L * s_L) <= tol_d
            converged = true; break
        end

        # Dual objective evaluation
        obj_y = dot(y_K, K_eff) + y_L * L_eff
        for i in 1:n
            sig = sigma_v[i]; al = alpha_v[i]; p = max(pi_vec[i], eps_val)
            if abs(sig - 1.0) < 1e-6
                obj_y += al * log(max(al, eps_val) / p) - al
            else
                obj_y += (sig / (1.0 - sig)) * al^(1.0/sig) * p^(1.0 - 1.0/sig)
            end
        end

        L_curr /= L_scale_dn
        local log_x_L_new
        while true
            log_x_K_new .= log_y_K .+ (eta_K / L_curr) .* grad_K
            log_x_L_new  = log_y_L  + (eta_L / L_curr)  * grad_L

            trial_K = exp.(log_x_K_new); trial_L = exp(log_x_L_new)
            pi_trial = BtT(trial_K) .+ trial_L .* l_tilde_v
            
            obj_trial = dot(trial_K, K_eff) + trial_L * L_eff
            for i in 1:n
                sig = sigma_v[i]; al = alpha_v[i]; p = max(pi_trial[i], eps_val)
                if abs(sig - 1.0) < 1e-6
                    obj_trial += al * log(max(al, eps_val) / p) - al
                else
                    obj_trial += (sig / (1.0 - sig)) * al^(1.0/sig) * p^(1.0 - 1.0/sig)
                end
            end

            if obj_trial <= obj_y || L_curr > 1e6; break; end
            L_curr *= L_scale_up
        end

        # Adaptive momentum
        if dot(log_y_K .- log_x_K, grad_K) + (log_y_L - log_x_L)*grad_L < 0
            tk_curr = 1.0
        end

        tk_next = (1.0 + sqrt(1.0 + 4.0 * tk_curr^2)) / 2.0
        beta_t  = (tk_curr - 1.0) / tk_next
        log_y_K .= log_x_K_new .+ beta_t .* (log_x_K_new .- log_x_K)
        log_y_L  = log_x_L_new  + beta_t  * (log_x_L_new  - log_x_L)
        log_x_K .= log_x_K_new; log_x_L = log_x_L_new
        tk_curr = tk_next
    end

    lam_K = exp.(log_x_K); lam_L = exp(log_x_L)
    pi_star = BtT(lam_K) .+ lam_L .* l_tilde_v
    C_star = (max.(alpha_v, 1e-15) ./ max.(pi_star, eps_val)).^(1.0 ./ sigma_v)
    X_star = neumann_apply!(cache, A_m, C_star .+ dK_v .+ G_v, k)

    return (C_star=C_star, X_star=X_star, pi_star=pi_star, success=converged, iterations=opt_iter)
end

# ── Synthetic data generator ─────────────────────────────────────────────────
function power_iteration(A::SparseMatrixCSC, max_iter=20)
    n = size(A, 1)
    b_k = rand(n)
    for _ in 1:max_iter
        b_k1 = A * b_k
        b_k = b_k1 ./ norm(b_k1)
    end
    return dot(b_k, A * b_k) / dot(b_k, b_k)
end

function make_synthetic(n::Int; sparsity=0.15, seed=42)
    Random.seed!(seed)

    # Sparse A with spectral radius ~ 0.60
    A_raw = sprand(n, n, sparsity)
    
    sr = abs(power_iteration(A_raw))
    A = A_raw * (0.60 / max(sr, 1e-10))

    delta = 0.0125
    B = rand(n) .* 0.25 .+ 0.05
    A_bar = A + delta * sparse(Diagonal(B))

    alpha = rand(n)
    alpha ./= sum(alpha)

    l_vec = rand(n) .* 0.4 .+ 0.1
    L_total = n * 0.3

    I_sparse = sparse(I, n, n)
    l_tilde = (I_sparse - A_bar)' \ l_vec

    X0 = rand(n) .* 2.0 .+ 1.0
    K = B .* X0 .* 1.2
    dK = rand(n) .* 0.04 .+ 0.01
    G = rand(n) .* 0.10 .+ 0.05
    
    sigma = rand(n) .* 0.5 .+ 0.75 # sigma in [0.75, 1.25]
    C_init = rand(n) .* 1.0 .+ 0.5

    return (alpha=alpha, A_bar=A_bar, B=B, l_tilde=l_tilde,
            dK=dK, K=K, L_total=L_total, G_vec=G, sigma=sigma, C_init=C_init)
end

# ── Benchmark ────────────────────────────────────────────────────────────────
# ── Final Benchmarking Loop & Plotting ────────────────────────────────────────
using Plots

function run_scaling_study()
    scales = 1000:1000:10000
    n_trials = 5
    avg_times = Float64[]

    println("Starting Scaling Study (CRRA Solver)...")
    println("Trials per scale: $n_trials")
    println("\nN Sectors | Avg Time (s)")
    println("-----------------------")

    for n in scales
        data = make_synthetic(n)
        # Warmup for this N
        solve_planner(data.alpha, data.A_bar, data.B, data.l_tilde, 
                      data.dK, data.K, data.L_total, data.G_vec, data.sigma, data.C_init)
        
        t_total = 0.0
        for _ in 1:n_trials
            t0 = time()
            solve_planner(data.alpha, data.A_bar, data.B, data.l_tilde, 
                          data.dK, data.K, data.L_total, data.G_vec, data.sigma, data.C_init)
            t_total += (time() - t0)
        end
        avg_t = t_total / n_trials
        push!(avg_times, avg_t)
        @printf("%-10d | %-12.4f\n", n, avg_t)
    end

    # Plotting
    p = plot(scales, avg_times, marker=(:circle, 5), lw=2,
             title="Dual Solver Complexity Scaling (CRRA)",
             xlabel="Number of Sectors (n)",
             ylabel="Wall-clock Time (s)",
             label="Avg Time (5 trials)",
             grid=true, legend=:topleft,
             color=:blue, dpi=300)
    
    # Quadratic Regression: y = a*n^2 + b*n + c
    # Form design matrix X
    X = [ones(length(scales)) collect(scales) collect(scales).^2]
    y = avg_times
    # Solve via normal equations: beta = (X'X) \ (X'y)
    beta = (X' * X) \ (X' * y)
    c, b, a = beta

    plot!(scales, c .+ b .* scales .+ a .* scales.^2, 
          ls=:dash, color=:orange, lw=2, label="Quadratic O(n^2) fit")

    @printf("\nQuadratic Fit: t(n) = %.2e n^2 + %.2e n + %.2e\n", a, b, c)

    out_path = "scaling_crra.png"
    savefig(p, out_path)
    println("Plot saved to $out_path")
end

if abspath(PROGRAM_FILE) == @__FILE__
    run_scaling_study()
end
